File: src/java/src/Generics.java
//: src/java/src/Generics.java
// {NewFeature} Preview in JDK 17
// Compile with javac flags:
//   --enable-preview --source 17
import java.util.List;

public class Generics {
  static <T> T nourish(T x) {
    // x.consume(); // Nope
    switch(x) {
      case Person p -> p.consume();
      case Robot r -> r.charge();
      default -> throw new IllegalStateException("Unexpected: " + x);
    }
    return x;
  }
  static Object nourish2(Object x) {
    switch(x) {
      case Person p -> p.consume();
      case Robot r -> r.charge();
      default -> throw new IllegalStateException("Unexpected: " + x);
    }
    return x;
  }
  //---
  public static void main(String... args) {
    List.of(new Person(), new Robot()).forEach(Generics::nourish);
    Person p = nourish(new Person());
    Object p2 = nourish2(new Person());
  }
}

File: src/kotlin/src/main/kotlin/Generics.kt
//: src/kotlin/src/main/kotlin/Generics.kt

fun <T> nourish(subject: T) : T =
    when (subject) {
        is Dog -> {
            subject.gulp()
            subject.bark()
            subject
        }
        is Person -> {
            subject.scarf()
            subject.greet()
            subject
        }
        is Robot -> {
            subject.charge()
            subject.initiate()
            subject
        }
        is Slug -> {
            subject.absorb()
            subject
        }
        else -> subject
    }

fun main() {
    listOf(
        Dog(), Person(), Robot(), Slug()
    ).map { nourish(it) }.forEach{ println(it) }
    val d: Dog = nourish(Dog())
    println(d)
}
